File: Structural/Lab12/Typescript/src/Entities/apartment.ts
import { RealEstate } from './real_estate';
import { Customer } from './customer';
import { Realtor } from './realtor';
import { RealEstateFlyweight } from './real_state_flyweight';
export class Apartment extends RealEstate {
    public roomsNumber: number;
    public heatingType: string;

    constructor();
    constructor(address: string, description: string, area: number, price: number, status: string, roomsNumber: number);
    constructor(address: string, description: string, area: number, price: number, status: string, roomsNumber: number, intrinsicState: RealEstateFlyweight);
    constructor(address?: string, description?: string, area?: number, price?: number, status?: string, roomsNumber?: number, intrinsicState?: RealEstateFlyweight)
    {
        if (address && description && area && price && status && roomsNumber) {
            super(address, description, area, price, status);
            if(intrinsicState){
                this.intrinsicState = intrinsicState;
            }
            this.roomsNumber = roomsNumber;
        } else {
            super("Dungeon", "tra ta ta", 50, 1000, "New");
            this.roomsNumber = 3;
        }
    }

    public toString(): string {
        return "Apartment:\n" + super.toString() + `, Rooms number: ${this.roomsNumber}`;
    }

    public rent(customer: Customer, startDate: Date, endDate: Date): void {
        if (this.isRentable)
            console.log(`Customer ${customer.nameAndSurname} rented apartment on address ${this.address} from ${startDate} to ${endDate}.`);
    }

    public sell(customer: Customer): void {
        if (this.isSellable)
            console.log(`Customer ${customer.nameAndSurname} bought apartment on address ${this.address}.`);
    }

    public setEstateToBuy(customer: Customer, price: number): void {
        if (this.isBuyable) {
            console.log(`Customer ${customer.nameAndSurname} put up for sale apartment on address ${this.address} with price(USD) ${price}.`);
            this.priceInUSD = price;
            this.isBuyable = false;
        }
    }

    public clone(): Apartment {
        const clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        clone.realtor = this.realtor ? Realtor.newRealtorFromObject(this.realtor) : null;
        return clone;
    }

    public shallowCopy(): Apartment {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    public reset(): void {
        super.reset();
        this.roomsNumber = 0;
        this.heatingType = "";
    }
}

File: Creational/Lab6/Python/apartment.py
from real_estate import *

class Apartment(RealEstate):
    def __init__(self, address="Dungeon", description="tra ta ta", area=50, price=1000, status="New", rooms_number=3):
        super().__init__(address,description,area,price,status)
        self.rooms_number = rooms_number
        self.realtor = None
        self.heating_type = None

    def __str__(self):
        return f"Apartment:\n{super().__str__()}, Rooms number: {self.rooms_number}"

    def rent(self, customer, start_date, end_date):
        if self.is_rentable:
            print(f"Customer {customer.name_and_surname} rented apartment on address {self.address} from {start_date} to {end_date}.")

    def sell(self, customer):
        if self.is_sellable:
            print(f"Customer {customer.name_and_surname} bought apartment on address {self.address}.")

    def buy(self, customer):
        if self.is_buyable:
            print(f"Customer {customer.name_and_surname} sold apartment on address {self.address}.")
    
    def clone(self):
        clone = copy.deepcopy(self)
        clone.realtor = None if clone.realtor is None else Realtor( self.realtor.name_and_surname, self.realtor.phone, self.realtor.email)
        return clone

    def shallow_copy(self):
        return copy.copy(self)
    
    def reset(self):
        super().reset()
        self.rooms_number = 0
        self.heating_type = None